src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+ *  child exits, but all other windows (if any) will receive events. Otherwise it'll return,
+ *  and it's your responsibility to keep @a stdout_text, @a stderr_text and @a exit_status
+ *  valid until the execution finishes.
+ *
+ *  @a stdin_text will be gfree()-d automatically when it's sent entirely to the child, or
+ *  when the child exits.
+ *
+ *  @a stdout_text and @a stderr_text will be set when the child exits.
+ *
+ *  @param working_directory child's current working directory, or @c NULL.
+ *  @param command_line child program and arguments, or @c NULL.
+ *  @param argv child's argument vector, or @c NULL.
+ *  @param envp child's environment, or @c NULL.
+ *  @param block_window window whose message loop to block until the child exits, or @c NULL.
+ *  @param stdin_text text to send to childs's stdin, or @c NULL.
+ *  @param stdout_text return location for the child's stdout, or NULL.
+ *  @param stderr_text return location for the child's stderr, or NULL.
+ *  @param exit_status return location for the child exit code, or NULL.
+ *  @param error return location for error.
+ *
+ *  @return @c TRUE on success, @c FALSE if an error was set.
+ **/
+gboolean spawn_with_capture(const gchar *working_directory, const gchar *command_line,
+	gchar **argv, gchar **envp, GtkWindow *block_window, gchar *stdin_text,
+	gchar **stdout_text, gchar **stderr_text, gint *exit_status, GError **error)
+{
+	SpawnCapture *sp = g_new(SpawnCapture, 1);
+
+	sp->stdout_text = stdout_text;
+	sp->stdout_str = stdout_text ? g_string_sized_new(1024) : NULL;
+	sp->stderr_text = stderr_text;
+	sp->stderr_str = stderr_text ? g_string_sized_new(1024) : NULL;
+	sp->exit_status = exit_status;
+
+	return spawn_with_callbacks(working_directory, command_line, argv, envp, block_window,
+		stdin_text, stdout_text ? spawn_capture_text : NULL, sp->stdout_str,
+		stderr_text ? spawn_capture_text : NULL, sp->stderr_str, spawn_capture_exit, sp,
+		NULL, error);
+}
+
+
+#ifdef SPAWN_TEST
+#include <stdio.h>
+#include <stdlib.h>
+
+#ifdef G_OS_UNIX
+#include <sys/types.h>
+#include <sys/wait.h>
+#endif
+
+
+static const char *read_line(const char *prompt, char *buffer, size_t size)
+{
+	fputs(prompt, stdout);
+
+	if ((buffer = fgets(buffer, size, stdin)) != NULL)
+	{
+		char *s = strchr(buffer, '\n');
+
+		if (*s)
+			*s = '\0';
+	}
+
+	return buffer;
+}
+
+
+static void printer(const gchar *s, gpointer user_data)
+{
+	printf("%s: %s", user_data == stdout ? "stdout" : "stderr", s);
+}
+
+
+static void exit_command(gint status, gpointer user_data)
+{
+	printf("finished, full exit code %d\n", status);
+	g_main_loop_quit((GMainLoop *) user_data);
+}
+
+
+typedef struct _CaptureData
+{
+	gchar *stdout_text;
+	GMainLoop *loop;
+} CaptureData;
+
+
+static gboolean capture(gpointer user_data)
+{
+	CaptureData *cd = (CaptureData *) user_data;
+
+	if (cd->stdout_text)
+	{
+		g_main_loop_quit(cd->loop);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+
+int main(int argc, char **argv)
+{

```

Comment:
A small testing program for various levels of execution. I can add a better "usage"...